service/resource/dnsrecord/resource.go

```diff
@@ -0,0 +1,351 @@
+}
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	o, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r.getCurrentState(ctx, o)
+}
+
+func (r *Resource) getCurrentState(ctx context.Context, obj azuretpr.CustomObject) (dnsRecords, error) {
+	recordSetsClient, err := r.getDNSRecordSetsClient()
+	if err != nil {
+		return nil, microerror.Maskf(err, "retrieving current state")
+	}
+
+	current := newPartialDNSRecords(obj)
+
+	for i, record := range current {
+		resp, err := recordSetsClient.Get(record.ZoneRG, record.Zone, record.RelativeName, dns.NS)
+		if client.ResponseWasNotFound(resp.Response) {
+			continue
+		} else if err != nil {
+			return nil, microerror.Maskf(err, "retrieving current state: getting record=%#v", record)
+		}
+
+		var nameServers []string
+		for _, ns := range *resp.NsRecords {
+			nameServers = append(nameServers, *ns.Nsdname)
+		}
+
+		current[i].NameServers = nameServers
+	}
+
+	return current, nil
+}
+
+// GetDesiredState returns the desired resource group for this cluster.
+func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
+	o, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Maskf(err, "computing desired state")
+	}
+
+	return r.getDesiredState(ctx, o)
+}
+
+func (r *Resource) getDesiredState(ctx context.Context, obj azuretpr.CustomObject) (dnsRecords, error) {
+	zonesClient, err := r.getDNSZonesClient()
+	if err != nil {
+		return nil, microerror.Maskf(err, "GetDesiredState")
+	}
+
+	desired := newPartialDNSRecords(obj)
+
+	for i, record := range desired {
+		zone := record.RelativeName + "." + record.Zone
+		resp, err := zonesClient.Get(key.ResourceGroupName(obj), zone)
+		if client.ResponseWasNotFound(resp.Response) {
+			return dnsRecords{}, nil
+		} else if err != nil {
+			return nil, microerror.Maskf(err, "GetDesiredState: getting zone=%q", zone)
+		}
+
+		var nameServers []string
+		for _, ns := range *resp.NameServers {
+			nameServers = append(nameServers, ns)
+		}
+
+		desired[i].NameServers = nameServers
+	}
+
+	return desired, nil
+}
+
+// NewUpdatePatch returns the patch creating resource group for this cluster if
+// it is needed.
+func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
+	o, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Maskf(err, "NewUpdatePatch")
+	}
+	c, err := toDNSRecords(currentState)
+	if err != nil {
+		return nil, microerror.Maskf(err, "NewUpdatePatch")
+	}
+	d, err := toDNSRecords(desiredState)
+	if err != nil {
+		return nil, microerror.Maskf(err, "NewUpdatePatch")
+	}
+
+	patch := framework.NewPatch()
+
+	updateChange, err := r.newUpdateChange(ctx, o, c, d)
+	if err != nil {
+		return nil, microerror.Maskf(err, "NewUpdatePatch")
+	}
+
+	patch.SetUpdateChange(updateChange)
+	return patch, nil
+}
```

Comment:
We have a more simplified pattern with this function in other operators as well.

Reply:
I extracted more private methods. E.g. for `GetDesiredState` I created a method `getDesiredState` which takes proper types. I think this allows to skip the boilerplate and focus on important parts when reading the code.
Natural consequence of that was to make `newUpdateChange` properly typed so I did conversions here. I could extract `newUpdatePatch` so it'd be more aligned maybe. WDYT?

Reply:
AFAIK we already talked about keeping `New*Patch` methods dumb to only gather the required changes. Adding specific type assertions here makes them more unique and always different which makes potential additional abstractions and human reading harder. All logic and type assertion happens in the `new*Change` methods so far and this project here should not differ out of a sudden. We can talk about how to do it in general but I would prefer keeping it consistent across the projects since this is a very powerful property of our projects in general.